adafruit_httpserver/server.py

```diff
@@ -0,0 +1,200 @@
+
+        :param socket: An object that is a source of sockets. This could be a `socketpool`
+          in CircuitPython or the `socket` module in CPython.
+        """
+        self._buffer = bytearray(1024)
+        self._timeout = 0
+        self.route_handlers = {}
+        self._socket_source = socket_source
+        self._sock = None
+        self.root_path = "/"
+
+    def route(self, path: str, method: HTTPMethod = HTTPMethod.GET):
+        """Decorator used to add a route.
+
+        :param str path: filename path
+        :param HTTPMethod method: HTTP method: HTTPMethod.GET, HTTPMethod.POST, etc.
+
+        Example::
+
+            @server.route(path, method)
+            def route_func(request):
+                raw_text = request.raw_request.decode("utf8")
+                print("Received a request of length", len(raw_text), "bytes")
+                return HTTPResponse(body="hello world")
+
+        """
+
+        def route_decorator(func: Callable) -> Callable:
+            self.route_handlers[_HTTPRoute(path, method)] = func
+            return func
+
+        return route_decorator
+
+    def serve_forever(self, host: str, port: int = 80, root_path: str = "") -> None:
+        """Wait for HTTP requests at the given host and port. Does not return.
+
+        :param str host: host name or IP address
+        :param int port: port
+        :param str root: root directory to serve files from
+        """
+        self.start(host, port, root_path)
+
+        while True:
+            try:
+                self.poll()
+            except OSError:
+                continue
+
+    def start(self, host: str, port: int = 80, root_path: str = "") -> None:
+        """
+        Start the HTTP server at the given host and port. Requires calling
+        poll() in a while loop to handle incoming requests.
+
+        :param str host: host name or IP address
+        :param int port: port
+        :param str root: root directory to serve files from
+        """
+        self.root_path = root_path
+
+        self._sock = self._socket_source.socket(
+            self._socket_source.AF_INET, self._socket_source.SOCK_STREAM
+        )
+        self._sock.bind((host, port))
+        self._sock.listen(10)
+        self._sock.setblocking(False)  # non-blocking socket
+
+    def poll(self):
+        """
+        Call this method inside your main event loop to get the server to
+        check for new incoming client requests. When a request comes in,
+        the application callable will be invoked.
+        """
+        try:
+            conn, _ = self._sock.accept()
+            with conn:
+                conn.settimeout(self._timeout)
+                received_data = bytearray()
+
+                # Receiving data until timeout
+                while "Receiving data":
+                    try:
+                        length = conn.recv_into(self._buffer)
+                        received_data += self._buffer[:length]
```

Comment:
In the blocking case, we need a way to know when the request is complete?

Reply:
Realized there is no blocking case currently allowed (timeout must be >=0, not `None`), but perhaps there should be?

Reply:
I agree that giving an option to set timeout to `None` (blocking) should be possible, thus not liming if one wants to do so.

It might also be logical to increase default timeout to `0.5` or `1` that would allow receiving request without explicitly setting timeout using `server.socket_timeout`.

Reply:
To be honest I do not really see any other way of determining whether it is the end of mesage other than timeout, I do not like that it unnecessarily waits after last part of message is received but it is the only way that seems to work no matter which client wants to connect to server. But, compared to current `main` branch, each request takes at least 2x longer.

One positive here is that it makes it possible to send data longer that max buffer size of 2880 bytes, which makes it more universal.

It is hard to find a middle point here, and we have to remember that we are talking about microcontrollers, not load-balanced web servers, so maybe a longer delay is not as much of a problem as the buffer size limit.